File: Assembly/ReportDocumentTree/vb/ReportDocumentTree/DocumentItem.vb
Imports Newtonsoft.Json
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Public Class DocumentItem
   Public Sub New()
        DocumentItems = New List(Of DocumentItem)()
   End Sub

    Public Sub New(ByVal occurrence As SolidEdgeAssembly.Occurrence)
        Me.New()
        FileName = occurrence.OccurrenceFileName
    End Sub

    Public Sub New(ByVal subOccurrence As SolidEdgeAssembly.SubOccurrence)
        Me.New()
        FileName = subOccurrence.SubOccurrenceFileName
    End Sub

    Public Property FileName() As String

    ''' <summary>
    ''' Returns all direct occurrences.
    ''' </summary>
    Public Property DocumentItems() As List(Of DocumentItem)

    ''' <summary>
    ''' Returns all direct and descendant children.
    ''' </summary>
'INSTANT VB TODO TASK: VB does not support iterators prior to VB11 (VS 2012):
    <JsonIgnore> _
    Public ReadOnly Iterator Property AllDocumentItems() As IEnumerable(Of DocumentItem)
        Get
            For Each bomItem In DocumentItems
'INSTANT VB TODO TASK: VB does not support iterators prior to VB11 (VS 2012):
                Yield bomItem

                For Each childBomItem In bomItem.AllDocumentItems
'INSTANT VB TODO TASK: VB does not support iterators prior to VB11 (VS 2012):
                    Yield childBomItem
                Next childBomItem
            Next bomItem
        End Get
    End Property

    Public Overrides Function GetHashCode() As Integer
        Return FileName.GetHashCode()
    End Function

    Public Overrides Function Equals(ByVal obj As Object) As Boolean
        If obj Is Nothing Then
            Return False
        End If

        If TypeOf obj Is DocumentItem Then
            Return Equals(DirectCast(obj, DocumentItem))
        End If

        Return MyBase.Equals(obj)
    End Function

    Public Overloads Function Equals(ByVal obj As DocumentItem) As Boolean
        If obj Is Nothing Then
            Return False
        End If

        Return obj.FileName.Equals(FileName)
    End Function

    Public Overrides Function ToString() As String
        Return FileName
    End Function
End Class

File: Assembly/ReportDocumentTree/cs/ReportDocumentTree/DocumentItem.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ReportDocumentTree
{
    public class DocumentItem
    {
       public DocumentItem()
        {
            DocumentItems = new List<DocumentItem>();
        }

        public DocumentItem(SolidEdgeAssembly.Occurrence occurrence)
            : this()
        {
            FileName = occurrence.OccurrenceFileName;
        }

        public DocumentItem(SolidEdgeAssembly.SubOccurrence subOccurrence)
            : this()
        {
            FileName = subOccurrence.SubOccurrenceFileName;
        }

        public string FileName { get; set; }

        /// <summary>
        /// Returns all direct occurrences.
        /// </summary>
        public List<DocumentItem> DocumentItems { get; set; }

        /// <summary>
        /// Returns all direct and descendant children.
        /// </summary>
        [JsonIgnore]
        public IEnumerable<DocumentItem> AllDocumentItems
        {
            get
            {
                foreach (var bomItem in DocumentItems)
                {
                    yield return bomItem;

                    foreach (var childBomItem in bomItem.AllDocumentItems)
                    {
                        yield return childBomItem;
                    }
                }
            }
        }

        public override int GetHashCode()
        {
            return FileName.GetHashCode();
        }

        public override bool Equals(object obj)
        {
            if (obj == null) return false;
            
            if (obj is DocumentItem)
            {
                return Equals((DocumentItem)obj);
            }

            return base.Equals(obj);
        }

        public bool Equals(DocumentItem obj)
        {
            if (obj == null) return false;

            return obj.FileName.Equals(FileName);
        }

        public override string ToString()
        {
            return FileName;
        }
    }
}
